### ISSUE TITLE
GC queue clogs SQL throughput

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#15756]
Folks on RC1 have reported a drop in SQL performance that correlates with GC queue activity (and vanishes, at least temporarily, when the GC queue policy is relaxed to 1 week):

![pasted image at 2017_05_08 06_37 pm](https://cloud.githubusercontent.com/assets/5076964/25801676/a4aa974a-33bc-11e7-82cb-c2c6ba165374.png)
![pasted image at 2017_05_08 06_35 pm](https://cloud.githubusercontent.com/assets/5076964/25801677/a4afc742-33bc-11e7-9194-dc2e8101624d.png)

The GC queue code leaves a lot to be desired. Currently, it piles up a potentially big list of keys; it may issue a lot of pushes/intent resolutions; proposer-evaluated KV possibly adds some extra pitfalls since for each GC request by adding some extra blow-up in size of the proposals.

Action items here:

- [ ] collect statistics. They are now guarded by `V(1)` and thus never available when we need them. Make available from SQL? Graph? Whatever the choice, they should be readily available. Filed https://github.com/cockroachdb/cockroach/issues/15769.
- [x] add this scenario to our load generators (https://github.com/cockroachlabs/production/issues/378), and/or:
- [x] ~~reproduce. We can seed a data set that will immediately need a lot of GC'ing without problems. Take motivation from the real world examples we see.~~ (see below for one example).
- [x] improve performance, starting with the area that needs most improvement. My hunch is that we're generating large proposals, but no proof at this time.
- [ ] GC should only block affected keys in the command queue. It blocks "everything" now: #7880

**Comment by petermattis** [comment]
Also, we should extend `kv` (and start using it on test clusters) to issue updates and deletes as well so that we're exercising the GC queue code path more frequently.

**Comment by tschottdorf** [comment]
@bdarnell pointed out:

> the problem is not necessarily the size of the GC. Any GC currently triggers a command-queue stall for its entire range

**Comment by tschottdorf** [comment]
Updated the action items to reflect command-queue and load generators.

**Comment by tschottdorf** [comment]
Looking into the `debug zip` log from the affected cluster, and in particular range 51 due to the following recurring error (still after bumping the GC ttl):

```
W170507 22:08:13.485520 131 storage/replica.go:2273  [gc,n5,s5,r6585/2:/Table/51/4/"270{2575[3m<80>-7765[3m<80>}] context cancellation after 248.5s of attempting command GC [/Table/51/4/"2702575281"/NULL/"4100911234389132",/Table/51/4/"2707765025"/1922-09-29T11:38:59.999999999Z/"4092509082411901")
```

The range is sizable, ~660k keys and a gc bytes age that would trigger the standard GC ttl, corroborating the fact that increasing the TTL helped the cluster (temporarily, until the data ages more).

This means that as assumed initially, we're likely looking at a performance problem once the GC queue has to deal with larger amounts of data. Furthermore, since the log message at the top is `DistSender` warning us that the `GC` batch never returns, we can also assume that it's not pushing/intent resolution that hangs; it's the actual garbage collection request (which is sent only at the end).

Now to figure out where it's stuck. My theory is that it's too large to apply (similar to a large range deletion) and thus is stuck in the command queue, where it blocks all other commands. What's at odds with this theory is that the cluster is presumably working now, and we still see the `DistSender` messages, so the request is still being sent to the range.

That leads to another theory: the GC queue timer duration is very low, one second. This was made so back in 2015 in 19a2bb71c7bf409e438dcf9ed03dfd0d85bc35a2 when the GC queue was split out of the (now nonexistent) Scan queue, which has a much lower interval.
This could conceivably compound the problem: if the GC request doesn't make it through and the GC queue gives up (even though the request might still be running around), the queue could promptly process the replica again, adding another request, ad absurdum. That would explain why the pressure eased up after relaxing the TTL. Anyway, those are no more than theories at this point.

**Comment by tschottdorf** [comment]
Actually, looking at the graph above, it seems that there are a lot of GC queue runs which then sharply drop. That does seem compatible: we pile them up and things get worse until we get lucky and one makes it through (or the TTL is increased and pressure is eased that way).

**Comment by tschottdorf** [comment]
I was able to reproduce badness. Not sure if it's exactly the same badness, but that can be investigated.

**Comment by a6802739** [comment]
@tschottdorf  about this `My theory is that it's too large to apply (similar to a large range deletion) and thus is stuck in the command queue, where it blocks all other commands`, how could this happen?

I mean what do you mean for this `it's too large to apply`? will CDB limit the apply data's size?

And why GC will block other commands?

**Comment by tschottdorf** [comment]
@a6802739 Similar to https://github.com/cockroachdb/cockroach/issues/15723, the GC queue can generate large proposals (if there are a lot of GCable keys in the range). We need to add some limitation there and/or make what's being sent over the wire smaller.

**Comment by a6802739** [comment]
if  the Raft leadership gets lost when these proposals sent over the network, The new raft leader will try to send these these proposals, right? Or it just response error to the client?

And I saw  `adding range tombstones at the MVCC level to speed up DROP TABLE` from #15723, any reference for this?

Thank you very much!

**Comment by petermattis** [comment]
@a6802739 No, we don't have another reference for the range tombstones idea yet. It has been kicking around internally for a bit and we're planning to implement it for 1.1. The range tombstones wouldn't help with large proposals due to GC, but I'd imagine we can limit the size of commands that GC generates.

**Comment by tschottdorf** [comment]
> if the Raft leadership gets lost when these proposals sent over the network, The new raft leader will try to send these these proposals, right? Or it just response error to the client?

Correct, in the absence of a node crash you normally shouldn't receive these errors at the client.

**Comment by a6802739** [comment]
@petermattis, Thank you very much.

@tschottdorf, the leadership lost because of blocking heartbeat to follower, right? So when the follower start a new election, And  if it's raft log didn't have this proposal, what will happen then?

**Comment by bdarnell** [comment]
> And if it's raft log didn't have this proposal, what will happen then?

It depends on the order of messages. Suppose A is the leader and proposes this large entry, but B and C do not have it. When B starts a new election, if it gets a response from A first, the election will fail (A will not vote for a node whose log is behind), and A will start a new election. If it gets a response from C first, the election will succeed and when B becomes the leader it will tell A to discard its proposal.

**Comment by a6802739** [comment]
if the proposal is discarded, the client will notice this?

And if B or C have the large entry, it will not start the new election, right?

And if it gets a response from A first, why the election will fail(doesn't it will wait the election from C), and why will it cause A start a new election?

**Comment by tschottdorf** [comment]
My local GC queue badness is going in a different direction than expected. I am on a single cluster, so Raft is less crucial in my scenario. However, what I'm seeing has nothing to do with it anyways.

1. local tracing shows that the GC queue indeed processes the same range (13 in the pastes below) over and over, and that there are "context exceeded" errors.
1. however, many times the GC actually goes through. Yet it is immediately queued again!

```
2017/05/09 10:36:05.781432	72.522447	gc
10:37:18.302951	72.521518	... [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] read-write path
10:37:18.302967	 .    16	... [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] command queue
10:37:18.303766	 .   799	... storage/replica.go:1686 [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] context deadline exceeded before command queue: GC [/Table/51/1/-1165984355386537541,/Max)
10:37:18.303811	 .    45	... [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] context deadline exceeded before command queue: GC [/Table/51/1/-1165984355386537541,/Max)
10:37:18.303831	 .    20	... [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] replica.Send got error: context deadline exceeded
10:37:18.303848	 .    17	... [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] context deadline exceeded
```

```
I170509 11:41:16.318564 33 storage/gc_queue.go:148  [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] gc queue: gcscore = 377.857971, intentscore = 0.000000, stats={ContainsEstimates:false LastUpdateNanos:1494340249843923900 IntentAge:0 GCBytesAge:2880980773362 LiveBytes:1503204 LiveCount:1 KeyBytes:56821 KeyCount:1 ValBytes:7418480961 ValCount:4734 IntentBytes:0 IntentCount:0 SysBytes:536 SysCount:9}
```

`GCBytesAge` slowly creeps up over time. That suggests that perhaps we're not keeping stats correctly. You certainly don't expect a range to be eligible for GC just after having been GC'ed - the GCBytesAge should be close to zero (or even zero right when the range is blank).
I manually issued a `DELETE FROM test.kv` and the problems continued (note `LiveCount` and `LiveBytes` decreased):

```
I170509 11:57:19.426329 118 storage/gc_queue.go:148  [gc,n1,s1,r13/1:/{Table/51/1/-…-Max}] gc queue: gcscore = 456.783416, intentscore = 0.000000, stats={ContainsEstimates:false LastUpdateNanos:1494344583505997193 IntentAge:0 GCBytesAge:35032923464208 LiveBytes:0 LiveCount:0 KeyBytes:56833 KeyCount:1 ValBytes:7418480961 ValCount:4735 IntentBytes:0 IntentCount:0 SysBytes:536 SysCount:9}
```

The last stats indicate that there's one key with only deleted versions, and 4735 of them. The size of that is ~ 7.4gb. No wonder iterating through the values alone takes the GC queue 72 seconds, as indicated above! To be fair, I wrote 1-2mb values, and this is coherent with what we're seeing: 1.5mb ~ 7418480961/4735 bytes in mb.

I'll take a look at why the gc bytes age can be so high but GC runs still leave that value alone. Should be straightforward, if tedious, from this point on.

**Comment by tschottdorf** [comment]
Looking at it now, the badness above makes perfect sense (though it's still badness).

An individual version will never be deleted unless it's 24 hours old (the default TTL). Nevertheless, each old version contributes to the GC age. We're really conflating to things here: we GC based on a fixed TTL (i.e. real_age) yet compute priority based on the GC age, which is real_age*size. In my case, `size >> 1`, and so, badness. This mismatch will be less problematic when `size ~ 1` for most values, but still I expect this to be problematic.

**Comment by tschottdorf** [comment]
We're definitely re-handling the queue on a context exceeded timeout right away (from trace inspection). That doesn't seem good. The queue timing in general seems very aggressive.

While I was at it, I verified that we're not leaking GC bytes age. After setting the TTL to 600 (which makes all keys eligible for deletion) the key and value count drop to zero along with `GCBytesAge` once the next successful GC runs:

`{ContainsEstimates:false LastUpdateNanos:1494348762964019048 IntentAge:0 GCBytesAge:0 LiveBytes:0 LiveCount:0 KeyBytes:0 KeyCount:0 ValBytes:0 ValCount:0 IntentBytes:0 IntentCount:0 SysBytes:536 SysCount:9}`

Definitely some work here to do and we may not even have scratched songhao's issue (though it might well be the mismatch between GCAge and age). I'll break it down in the initial post.

**Comment by tschottdorf** [comment]
Now working on our test cluster `adriatic`, running the `kv` loadgen with up to 1kb keys and 10k cycle length):

![image](https://cloud.githubusercontent.com/assets/5076964/25868202/e0c54000-34c9-11e7-9ec3-1eedcb6c278f.png)
![image](https://cloud.githubusercontent.com/assets/5076964/25868306/41d862a0-34ca-11e7-9e1e-fbd6f51d484b.png)

![image](https://cloud.githubusercontent.com/assets/5076964/25868267/184b348a-34ca-11e7-921f-90a06bf9743c.png)

We're seeing regular queue activity; 7s is ~median. Most of the time, preprocessing the key span takes the majority of time, though the time during which the GCRequest is in the command queue (i.e. blocks reads) is too large (1-2s).

Then, there are also rare pathological ones like this:

![image](https://cloud.githubusercontent.com/assets/5076964/25868432/a8b3f7d2-34ca-11e7-9032-1bbaefe73f3e.png)

Once https://github.com/cockroachdb/cockroach/pull/15812 is deployed, I should be able to pull more from the traces as to what's happening. My expectation is that we see the same problem here: GC runs, but none of the data written is 24h old. So it runs again, and again, and again; and that the generated Raft proposals are thus empty.

**Comment by tschottdorf** [comment]
Yeah, pretty clear evidence that we're GC'ing ranges but not actually deleting anything (was kind of clear that was going to happen, but good to see it explicitly with a reasonable workload):

```
0s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] iterating through range
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] pushing up to 0 transactions (concurrency 25)
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] resolving up to 0 intents
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] processing abort cache
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] assembled GC keys, now proceeding to GC; stats {Now:1494419549.178907435,0 Policy:{TTLSeconds:86400} NumKeysAffected:0 IntentsConsidered:0 IntentTxns:0 TransactionSpanTotal:0 TransactionSpanGCAborted:0 TransactionSpanGCCommitted:0 TransactionSpanGCPending:0 TxnSpanGCThreshold:1494415949.178907435,0 AbortSpanTotal:0 AbortSpanConsidered:0 AbortSpanGCNum:0 PushTxn:0 ResolveTotal:0 ResolveSuccess:0 Threshold:1494333149.178907435,0}
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] read-write path
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] command queue
2.32s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] waiting for 2 overlapping requests
3.19s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] waited 868.666953ms for overlapping requests
3.19s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] raft
4.65s
[gc,n2,s2,r736/2:/Table/51/1/85{23974…-44132…}] applying command
```

![image](https://cloud.githubusercontent.com/assets/5076964/25899021/3d399524-355c-11e7-8648-c477f9f0b906.png)

**Comment by tschottdorf** [comment]
The traces also show fairly consistently that the GC requests, though empty, probably cause a decent hiccup for the writers: During the 868.666953ms this command waits for previous writes, everything is blocked. So special-casing the GC key seems to make sense, though I'd want to isolate the impact a bit more explicitly first.

**Comment by tschottdorf** [comment]
I think `cobalt` shows roughly the same here now:

![image](https://cloud.githubusercontent.com/assets/5076964/25918892/f779fb0c-359a-11e7-9524-f3f68021c626.png)

There's no good reason for that much GC queue activity given the photos/block_writer workload.

**Comment by bdarnell** [comment]
Maybe this will become moot after #15897 lands and we're no longer generating empty GC requests, but it would be easy to add an early exit to `gcQueue.process` if `gcKeys` is empty, so we avoid sending these empty GC requests at all. Or, if we want to still send the request to update the persisted last-GC time even when we're not deleting anything, we could a narrow range on `gcArgs.{Key,EndKey}` (I think that would be safe to do now without going all the way to #7880, although we should double-check that there are no subtle interactions with splits that require the GC to cover the entire range no matter what keys it is operating on)

**Comment by tschottdorf** [comment]
I filed #15989 about chunking. #7880 takes care of blocking the command queue, at least to the point where it's likely less interesting to do more. I'll close this when #15897 lands.

**Comment by tschottdorf** [comment]
@a6802739 let us know if you're still having problems and we'll open a new issue!

**Comment by a6802739** [comment]
@tschottdorf, got it. Thank you very much, I will have a see on it then.
